Adding support for <link> css inclusion

FabioFerrettiMoodys opened the pull request:
#### adding support for &lt;Link ref="stylesheet" url=""&gt; 


#### Adding support for <link> css inclusion

####  Read the content of a <link rel="stylesheet" href="..> and add it like a <style> node.

####  in order to do that, the base uri must be provided as the SvgDocument.BaseUri is set after the Open function.

FabioFerrettiMoodys commented:
here is why the test are important ;)

i got a error in Open of svgDocument : i forgot the else ;)

       public static T Open<T>(string path, Dictionary<string, string> entities, Uri baseUri = null) where T : SvgDocument, new()
        {
            if (string.IsNullOrEmpty(path))
            {
                throw new ArgumentNullException("path");
            }

            if (!File.Exists(path))
            {
                throw new FileNotFoundException("The specified document cannot be found.", path);
            }

            using (var stream = File.OpenRead(path))
            {
                var doc = Open<T>(stream, entities, baseUri == null ? new Uri(System.IO.Path.GetFullPath(path)) : baseUri);
                if (baseUri == null)
                    doc.BaseUri = new Uri(System.IO.Path.GetFullPath(path));
                else
                    doc.BaseUri = baseUri;
                return doc;
            }
        }

H1Gdev commented on Source/SvgDocument.cs:
```diff
@@ -260,32 +260,35 @@ private static bool ExceptionCaughtIsGdiPlusRelated(Exception e)
         /// <returns>An <see cref="SvgDocument"/> with the contents loaded.</returns>
         /// <exception cref="FileNotFoundException">The document at the specified <paramref name="path"/> cannot be found.</exception>
-        public static SvgDocument Open(string path)
+        public static SvgDocument Open(string path, Uri baseUri = null)
```
Adding `Open` argument doesn't seem to be active. (ref. #457)

FabioFerrettiMoodys replied:
i don't understand.

H1Gdev commented on Source/SvgDocument.cs:
```diff
@@ -435,6 +443,19 @@ private static T Open<T>(XmlReader reader) where T : SvgDocument, new()
+                                    var href = element.CustomAttributes["href"];
+                                    Uri uri = baseUri == null ? new Uri(href) : new Uri(baseUri, href);
+                                    var content = File.ReadAllText(uri.AbsolutePath);
+                                    SvgUnknownElement elem = new SvgUnknownElement("style");
```
use `var`.

mrbean-bremen commented:
@H1Gdev - are there open review issues here?

H1Gdev commented:
There is no `<link>` element in the SVG specifications.

- [SVG](https://www.w3.org/TR/SVG11/eltindex.html)
- [SVG2](https://www.w3.org/TR/SVG2/eltindex.html)

@FabioFerrettiMoodys 

Please add target SVG files to your test.
Please tell me supported software (major browser etc.).

FabioFerrettiMoodys commented:
Hello

https://www.w3.org/TR/SVG2/styling.html

```
6.3. External style sheets: the effect of the HTML ‘link’ element
An HTML ‘link’ element in an SVG document (that is, an element in the HTML namespace with local name "link") with its ‘rel’ attribute set to 'stylesheet' must be processed as defined in the HTML specification and cause external style sheets to be loaded and applied to the document. Such elements in HTML documents outside of an inline SVG fragment must also apply to the SVG content.
```

all major browser seems to support it 
i've test this sample with chrome, ie11 and firefox.
http://www.schepers.cc/svg/style/external-link-xhtml.svg

H1Gdev commented:
This PR has the following issues.
- File output by `Draw` is not supported.
- Do not support the case if `href` is specified in URL.

First, make the following changes.
- Create an element for \<link\>.
- `BaseUri` is a user-settable property, so please resolve `link` `href` later.

H1Gdev commented:
### FYI

Output SVG with this PR is as follows.
```svg
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg width="100%" height="100%" viewBox="0, 0, 50, 50" xhtml="http://www.w3.org/1999/xhtml" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xml="http://www.w3.org/XML/1998/namespace" version="1.1">
  <title>Styling Test</title>
  <desc>
    Tests linking to external stylesheet with "link" element in XHML namespace.  
    Seems to work in Firefox, Opera, and Safari, as of 18-01-2009
  </desc>
  <metadata>author: schepers, created: 18-01-2009</metadata>
  <link href="http://www.schepers.cc/svg/style/svgstyle.css" rel="stylesheet" type="text/css" />
  <circle cx="25" cy="25" r="20" />
</svg>
```

FabioFerrettiMoodys commented:
> * `BaseUri` is a user-settable property, so please resolve `link` `href` later.

i don't know how to do it, you collect all the styles in the function Open<T>(XmlReader reader... 
You create a "resulting css" named "cssTotal" and assign the styles to te elements at the end of these function : before we can set the BaseUri.
line 480 of svgDocument.cs
` var cssTotal = styles.Select((s) => s.Content).Aggregate((p, c) => p + Environment.NewLine + c);
`

H1Gdev commented:
@mrbean-bremen 
@FabioFerrettiMoodys 

By #90, `BaseUri` property can be set...
We plan to ensure backward compatibility as much as possible.

This is modification to review style application sequence by supporting the SVG2 specification.

FabioFerrettiMoodys commented:
@H1Gdev 

> By #90, `BaseUri` property can be set...
> We plan to ensure backward compatibility as much as possible.
> 
> This is modification to review style application sequence by supporting the SVG2 specification.

I'm not sure to follow you : it can be set but only after the Open. but the styling is done in the Open. (SvgDocument.cs line 507 : svgDocument?.FlushStyles(true); ).
 i can make a link element and resolve the href later, but if a user set it, it will not work as all the styling is done at the open of the file. (if it's a new element, i will also need to redo the aggregate of styles at the end of the Open function. As the aggregate of the ISvGNode.Content will not work anymore.)

ps: sorry but i'm struggling to understand your comments and what and how you want me to do. I don't know your library as deep as you. Can you be more specific ? 

for exemple what do you mean by "File output by Draw is not supported." 
```
var svg = SvgDocument.Open(Source);
var image = svg.Draw();
```
this code work, do you mean something else ?

I also don't understand your "Output SVG with this PR is as follows."
how do you generate it like this ? i didn't alter the svgdocument structure, i've only added the css content to the list of styles used for the styling at the end of the Open.

H1Gdev commented:
> for exemple what do you mean by "File output by Draw is not supported."

It's a typo.

File output by `Write` is not supported.

H1Gdev commented:
>  it can be set but only after the Open. but the styling is done in the Open. (SvgDocument.cs line 507 : svgDocument?.FlushStyles(true); ).

This is because it doesn't support \<link\>.
This is because we was able to solve style here.

FabioFerrettiMoodys commented:
As the link element wasn't in the svg namespace i've added a NonSvgElementAttribute so the SvgElementFactory can create NonSvgElement by refraction like the SvgElement.

i've assigned the BaseUri just after the creation of the svgdocument in the open function : the svglink use the OwnerDocument.BaseUri when the content of the link is requested. (svgLink.GetLinkContentAsText)

i've corrected the svgDocument.Write : it write the <link> element, but i don't know if it is what you want as the styles are already added on each elements (by the flushstyles function).

H1Gdev commented on Source/Document Structure/SvgLink.cs:
```diff
@@ -0,0 +1,107 @@
+
+         public override SvgElement DeepCopy()
+        {
+            return DeepCopy<SvgTitle>();
```
SvgLink

FabioFerrettiMoodys replied:
should be ok

H1Gdev commented on Source/Document Structure/SvgLink.cs:
```diff
@@ -0,0 +1,107 @@
+        [SvgAttribute("href")]
+        public string Href
+        {
+            get { return GetAttribute<string>("href", false,string.Empty); }
```
add space.
`false, string.Empty`

FabioFerrettiMoodys replied:
should be ok

H1Gdev commented on Source/Document Structure/SvgLink.cs:
```diff
@@ -0,0 +1,107 @@
+    {
+        public SvgLink()
+          :base("link")
+		{
```
Not use TAB.

FabioFerrettiMoodys replied:
should be ok

H1Gdev commented on Source/SvgDocument.cs:
```diff
@@ -392,6 +400,8 @@ private static T Open<T>(XmlReader reader) where T : SvgDocument, new()
                             else
                             {
                                 svgDocument = elementFactory.CreateDocument<T>(reader);
+                                if(baseUri!=null)
```
Add space.

H1Gdev commented on Source/SvgElementFactory.cs:
```diff
@@ -14,8 +14,9 @@ namespace Svg
     {
-        private List<ElementInfo> availableElements;
-        private Parser cssParser = new Parser();
+		private List<ElementInfo> availableElements;
```
What changed ?

FabioFerrettiMoodys replied:
another automatic tabs/space remplacement by my texteditor configuration at work. sorry.

H1Gdev commented on Source/Document Structure/SvgLink.cs:
```diff
@@ -0,0 +1,107 @@
+
+        public Stream GetLinkContentAsStream(RelativeValue rel = RelativeValue.Unknown)
+        {
+            if (Rel != rel && rel != RelativeValue.Unknown)
```
If you're doing a comparison with `Rel` property, why not use the value of `Rel` property value ??

FabioFerrettiMoodys replied:
it is to avoid as much as possible the use of the content with a incorrect rel type.

if you call GetLinkContentAsStream(stylesheet) on a link  with rel="icon" for exemple, you will not get the content and cannot use it as a stylesheet by mistake (and make the code crash).

assuming you are absolutely sure of what you do and what you have, you can just call GetLinkContentAsStream(), you will get the content independently of the rel 

 these technique avoid to check everywhere if the Rel is correct every where you use the link content, here it is done in only one place.

I can remove if if you prefer.

H1Gdev commented on Source/Document Structure/SvgLink.cs:
```diff
@@ -0,0 +1,107 @@
+            get { return GetAttribute<RelativeValue>("rel", false, RelativeValue.Unknown); }
+            set { Attributes["rel"] = value; }
+        }
+
```
There is one blank line.

H1Gdev commented on Source/Document Structure/SvgLink.cs:
```diff
@@ -0,0 +1,107 @@
+        {
+            return DeepCopy<SvgLink>();
+        }
+
```
There is one blank line.

H1Gdev commented on Source/Document Structure/SvgLink.cs:
```diff
@@ -0,0 +1,107 @@
+        public SvgLink()
+          : base("link")
+        {
+
```
No blank lines are needed.

H1Gdev commented on Source/Document Structure/SvgLink.cs:
```diff
@@ -0,0 +1,107 @@
+
+        public string GetLinkContentAsText(RelativeValue rel)
+        {
+            var stream = GetLinkContentAsStream(rel);
```
use `using`

H1Gdev commented on Source/Document Structure/SvgLink.cs:
```diff
@@ -0,0 +1,107 @@
+                return null;
+
+            string content = null;
+            using (StreamReader sr = new StreamReader(stream))
```
use `var`.

H1Gdev commented on Source/Document Structure/SvgLink.cs:
```diff
@@ -0,0 +1,107 @@
+                var httpRequest = WebRequest.Create(uri);
+
+                var webResponse = httpRequest.GetResponse();
+                var stream = webResponse.GetResponseStream();
```
Please call `Dispose` when exception occurs.

mrbean-bremen commented:
@H1Gdev - can you please check if the current state can be merged? You did some review that has been addressed (admittetly, this was quite some time ago, but I hadn't been active in this project since then).

kimsey0 commented:
(If #872 is merged and this pull request is resumed, it should check that `SvgDocument.ResolveExternalResources == true` before creating a `WebRequest` in `GetLinkContentAsStream`.)

paulushub commented:
@FabioFerrettiMoodys It looks like your `link` requirement is non-standard, or please provide some standard use cases.

paulushub commented:
@FabioFerrettiMoodys Sorry for the delay. Please update your project for review.

kimsey0 commented on Source/Document Structure/SvgLink.cs:
```diff
@@ -0,0 +1,111 @@
+                    uri = new Uri(OwnerDocument.BaseUri, uri);
+
+                // should work with http: and file: protocol urls
+                var httpRequest = WebRequest.Create(uri);
```
This should check that `SvgDocument.ResolveExternalResources` is true before creating a web request to avoid XSS vulnerabilities. See #873.

paulushub replied:
@kimsey0 Can you fix it by editing the PR source?

kimsey0 replied:
It's been too long since I worked on this project, so I can't really remember the details, nor how tests etc. are structured. I can see that #873 split `SvgDocument.ResolveExternalResources` into three properties: `ResolveExternalXmlEntites`, `ResolveExternalImages`, and `ResolveExternalElements`, probably to give more fine-grained control over which external resources are allowed to be resolved. You may want to introduce a new `ExternalType ResolveExternalStylesheets` property, somehow pass it in or make it available to this `GetLinkContentAsStream` method, and then trace a warning and return null or an empty string if `!ResolveExternalStylesheets.AllowsResolving(uri)`.

paulushub replied:
@kimsey0 Thank for the update. Just do not know the best way to handle these staled PR.
- Take over (clone PR reposition and fix issues), the original owner looses his/her contribution
- Close the PR and see how to reimplement the issue, a waste but may be worth it

mrbean-bremen replied:
- adapt/finish the original PR, so it will just have 2 authors

paulushub replied:
> * adapt/finish the original PR, so it will just have 2 authors

Never done, can you outline the procedure and give me some hints?

mrbean-bremen replied:
If you are asking about the git workflow, something like:
> git remote add FabioFerrettiMoodys https://github.com/FabioFerrettiMoodys/SVG
> git checkout FabioFerrettiMoodys/master (probably under some other name)
> git rebase origin/main -> resolve conflicts as usual
> git push --force-with-lease
(or the equivalents if you use a git GUI client(
- make your changes, commit, push...

Is this what you are asking?

paulushub replied:
> Is this what you are asking?

I think so, thanks - will look into it this weekend.

kimsey0 replied:
Depending on the policies of this project (the contribution guide doesn't say), you can also just avoid squashing the commits, leaving the original author for each one and making it clear who's done what.

mrbean-bremen replied:
> the contribution guide doesn't say

Something to be added...

The current (undocumented) convention is that commits are usually squashed and only rebased if the individual commits are independent changes to be preserved, and have sensible commit comments (which is not the case here). Though it is possible to rebase the PR respectively before the merge, if a rebase is wanted.
